common/dictionary-db/dictionary-db.ts

```diff
@@ -0,0 +1,1416 @@
+                });
+
+                progress.current += modifiedCardsBatch.size;
+                await this._updateBuildAnkiCacheProgress(
+                    buildId,
+                    trackBuildIdsToClear,
+                    progress,
+                    Array.from(modifiedTokens),
+                    statusUpdates,
+                    t
+                );
+            },
+            { batchSize: 100 } // Batch for memory usage and yomitan cache size
+        );
+    }
+
+    private async _saveTokensForDB(
+        profile: string,
+        trackStates: Map<number, TrackStateForDB>,
+        records: DictionaryTokenRecord[],
+        ankiCards: DictionaryAnkiCardRecord[],
+        modifiedCardsBatch: CardsForDB,
+        trackTokensMap: Map<
+            number,
+            Map<DictionaryTokenSource, Map<string, { lemmas: string[]; cardIds: Set<number> }>>
+        >,
+        modifiedTokens: Set<string>
+    ): Promise<void> {
+        for (const record of records) {
+            modifiedTokens.add(record.token);
+            for (const lemma of record.lemmas) modifiedTokens.add(lemma);
+        }
+        return this.db.transaction('rw', this.db.tokens, this.db.ankiCards, async () => {
+            await Promise.all([this._saveRecordBulk(records), this.db.ankiCards.bulkPut(ankiCards)]);
+            await this.db.tokens
+                .where('cardIds')
+                .anyOf(Array.from(modifiedCardsBatch.keys()))
+                .distinct()
+                .filter((r) => r.profile === profile)
+                .modify((record, ref) => {
+                    if (
+                        record.source !== DictionaryTokenSource.ANKI_WORD &&
+                        record.source !== DictionaryTokenSource.ANKI_SENTENCE
+                    ) {
+                        return;
+                    }
+                    if (!trackStates.has(record.track)) return;
+                    if (trackTokensMap.get(record.track)!.get(record.source)!.has(record.token)) return; // We want tokens that were not updated but refers to updated cards (e.g. field value changed, different tokens)
+                    const validCardIds = record.cardIds.filter((id) => !modifiedCardsBatch.has(id));
+                    if (!validCardIds.length) {
+                        modifiedTokens.add(record.token);
+                        for (const lemma of record.lemmas) modifiedTokens.add(lemma);
+                        delete (ref as any).value;
+                    } else if (validCardIds.length !== record.cardIds.length) {
+                        modifiedTokens.add(record.token);
+                        for (const lemma of record.lemmas) modifiedTokens.add(lemma);
+                        record.cardIds = validCardIds;
+                    }
+                });
+        });
+    }
+
+    private async _gatherModifiedTokens(profile: string, modifiedTokens: Set<string>) {
```

Comment:
Suggestion to comment that `modifiedTokens` is being expanded by this method with existing entries in the database